### ISSUE TITLE
cvxpy fails to install on Apple Silicon under Rosetta2 Homebrew Python

### ISSUE THREAD

**Issue by sunilshah** [cvxpy/cvxpy#1190]
**Describe the bug**

Python 3.9.0_5 installs fine and so do most pypi packages as recommended for Homebrew 2.6.0 using Rosetta2. 

```  
Python % arch -x86_64 pip3 install cvxpy
Collecting cvxpy
  Using cached cvxpy-1.1.7.tar.gz (1.0 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/opt/python@3.9/bin/python3.9 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/tmpwb8n3knm
       cwd: /private/var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/pip-install-7edffu8y/cvxpy_b036aee14ee549eb9ea539d7d269f653
  Complete output (22 lines):
  Traceback (most recent call last):
    File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 280, in <module>
      main()
    File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 263, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 114, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/private/var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/pip-build-env-dcf2qxbq/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 149, in get_requires_for_build_wheel
      return self._get_build_requires(
    File "/private/var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/pip-build-env-dcf2qxbq/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 130, in _get_build_requires
      self.run_setup()
    File "/private/var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/pip-build-env-dcf2qxbq/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 253, in run_setup
      super(_BuildMetaLegacyBackend,
    File "/private/var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/pip-build-env-dcf2qxbq/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 145, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 39, in <module>
      python_target = distutils.version.LooseVersion(
    File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/version.py", line 306, in __init__
      self.parse(vstring)
    File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/version.py", line 314, in parse
      components = [x for x in self.component_re.split(vstring)
  TypeError: expected string or bytes-like object
  ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/local/opt/python@3.9/bin/python3.9 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /var/folders/sj/0mpllhrn5dgbk_sjm6l34jl40000gn/T/tmpwb8n3knm Check the logs for full command output.

```
**To Reproduce**
A [minimal code example](https://stackoverflow.com/help/mcve) that reproduces the bug or unexpected behavior.
 
See above. 

**Expected behavior**

Normal pip3 install under Rosetta2

**Output**
If applicable, include program output. If reporting a program crash, please include the entire stack trace.

See above

**Version**
 - OS: Mac Big Sur 11.0.1
 - CVXPY Version: 1.1.17
 - Apple Mac-mini. 2020.

**Additional context**

Using Rosetta 2 (``` arch -x86_64 ```), cvxopt, scs and ecos install and run fine.  So does Jupyter notebook.  

https://brew.sh/2020/12/01/homebrew-2.6.0/

**Comment by angeris** [comment]
Oh, that seems bad.

I will take a look this weekend or early next week, since I'm finishing up a paper.

**Comment by rileyjmurray** [comment]
@sunilshah we aren't testing our pip installations against python 3.9 yet. Could you try Python 3.7?

As a separate note: [our conda-forge builds](https://github.com/conda-forge/cvxpy-feedstock#current-release-info) do not currently have [osx-arm64 as a target architecture](https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/).

**Comment by sunilshah** [comment]
@rileyjmurray,  python 3.9 on Intel machines with Big Sur does work well under Homebrew.  cvxpy and the solvers (osqp, scs, ecos, cvxopt) work fine on Intel machines with Python 3.9 (and 3.8).   

I do not use Anaconda because I need other non-pythonic software under Homebrew.   
I am not sure Anaconda works on Apple M1.   On M1, Big Sur ships with native (arm64) python 3.8.2 in the standard Homebrew location of /usr/local/bin.   However it does not have a functional pip.  On M1, Homebrew (2.6.0 Dec 1 ) recommends use of Rosetta 2 and python 3.9 ```arch -x86_64 python3```.  That worked fine on M1 for most packages I use, except cvxpy and osqp. Hence the post about ```arch -x86_64``` install. 

It is not clear if I can install and test python 3.7 on M1 mac safely.   In python 3.8 and 3.9 there are a number of improvements over 3.7.  I had to modify my code base to reflect those on my Intel machines.  It would be wasteful to revert those changes.  What would you like to know from 3.7 ? 

What are your plans to support python 3.8 and 3.9? What about Apple M1 under Rosetta 2 and native ?   

In the mean time, I can run cvxpy on Intel machines with python 3.8 and 3.9.  M1 is significantly faster than my Intel machines.  Even under Rosetta 2 I see significant speedups.

**Comment by rileyjmurray** [comment]
@sunilshah if cvxpy works on python 3.9 on your intel machine then that's fine. I asked about 3.7 just because I know for sure we test against it.

We will add continuous integration testing for python 3.8 and 3.9 over the next month. Support for Apple M1 will initially be limited since none of the CVXPY core developers have an M1 machine (correct me if I'm wrong @SteveDiamond @akshayka).

**Comment by sunilshah** [comment]
@rileyjmurray I will be happy to help with testing on my M1 machine with  python 3.9+ homebrew.   I  also run cvxpy on python 3.8.5 on Ubuntu LTS 20.04 on Amazon AWS Intel instances. 

I believe, you may be able to get M1 instances on Amazon AWS for CI early next year.

**Comment by rileyjmurray** [comment]
@sunilshah per #1202 we now run our normal build process against python 3.6 (windows) and 3.7 -- 3.9 (windows, osx, linux).

Our ability to build on M1 is going to be contingent on when TravisCI or AppVeyor provide support for that architecture.

**Comment by sunilshah** [comment]
@rileyjmurray   

Thanks for the update on 3.9.   TravisCI does install on Big Sur using Homebrew.  There are Big Sur bottles for both x86_64 and arm64.

**Comment by sunilshah** [comment]
@rileyjmurray

I can confirm that cvxpy now installs and runs correctly under Rosetta 2 on Apple M1. 

Thanks.

brew installed  version python 3.9.1_3 
cvxpy version 1.1.7
Big Sur 11.1

**Comment by isuruf** [comment]
> As a separate note: our conda-forge builds do not currently have osx-arm64 as a target architecture.

It is blocked on scs and osqp at https://github.com/conda-forge/scs-feedstock/pull/19 and https://github.com/conda-forge/osqp-feedstock/pull/26 and

**Comment by sunilshah** [comment]
I have homebrew installed python 3.9.6  on Apple M1 (Mac OS 11.5.1) native (arm64).  For numpy (1.12.1)  and scipy (1.7.0) , I had to use homebrew installed versions as building from source did not work on M1. 
 
I was able to install and run cvxpy on M1 in native mode for cvx vesion 1.1.13 using the command 

``` 
pip3 install --no-use-pep517 cvxpy  

```
This was to avoid cvxpy install forcing a fresh install of numpy and scipy and failing at it.  
However, with version 1.1.14, the install fails with 

```
ERROR: Disabling PEP 517 processing is invalid: project specifies a build backend of setuptools.build_meta in pyproject.toml
```


``` pip3 install -U cvxpy ```  fails as expected because of numpy and scipy installed under homebrew.   

``` pip3 install -U --no-deps cvxpy ``` also fails as it seems to ignore the ```--no-deps``` flag. 

How to do I get around this issue?

**Comment by SteveDiamond** [comment]
There are no wheels on M1?

**Comment by sunilshah** [comment]
There are wheels--but not for important  pypi packages.  Under Rosetta 2, there are  wheels for most pypi packages on my M1 machine.  

For native M1 (-arch arm64), many popular numerical packages require compilation from source, though many can be have wheels.  Numpy,  Scipy are difficult to install even from source using pip so hombrew versions are needed. I was able to install pandas from source using pip.

**Comment by ustunb** [comment]
Hello! I'm following up on this issue. 

I'd like to install cvxpy on a M1 Mac using pip, rather than conda-forge. Is there a way to do this? 

Currently, most of the proposed solutions in this thread lead to various bugs. This includes:

```
pip3 install cvxpy
pip3 install -U cvxpy
pip3 install -U --no-deps cvxpy
```

I am using Python 3.9.4 but happy to switch. 

Happy to post as a separate issue, or follow up on Discord.

**Comment by sunilshah** [comment]
> Hello! I'm following up on this issue.
> 
> I'd like to install cvxpy on a M1 Mac using pip, rather than conda-forge. Is there a way to do this?
> 
> Currently, most of the proposed solutions in this thread lead to various bugs. This includes:
> 
> ```
> pip3 install cvxpy
> pip3 install -U cvxpy
> pip3 install -U --no-deps cvxpy
> ```
> 
> I am using Python 3.9.4 but happy to switch.
> 
> Happy to post as a separate issue, or follow up on Discord.

Have you tried ``` pip3 install -U -no-use-pep517 cvxpy``` ?   I have installed version cvxpy 1.1.17 on my M1 under OS 12.0.1.  Python3 version 3.9.9 was installed using Homebrew.

**Comment by ustunb** [comment]
Thanks!

I just tried to install using `pip install -U --no-use-pep517 --no-cache-dir cvxpy` but run into an error.

I'm including the error logs below. Any chance you know what might be going on? If you managed to get it working, I feel like it might be something silly.



```
Requirement already satisfied: cvxpy in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (1.1.8)
Collecting cvxpy
  Downloading cvxpy-1.1.17.tar.gz (1.3 MB)
     |████████████████████████████████| 1.3 MB 2.6 MB/s            
  Preparing metadata (setup.py) ... done
Collecting osqp>=0.4.1
  Downloading osqp-0.6.2.post0.tar.gz (219 kB)
     |████████████████████████████████| 219 kB 12.8 MB/s            
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_a770c588033e4c6eb1f79635087d3d40/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_a770c588033e4c6eb1f79635087d3d40/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-pip-egg-info-lyttq8n1
       cwd: /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_a770c588033e4c6eb1f79635087d3d40/
  Complete output (50 lines):
  WARNING: The wheel package is not available.
    ERROR: Command errored out with exit status 1:
     command: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-vfugjp8k/qdldl_6df36527210846cdaee726a9ee2d998f/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-vfugjp8k/qdldl_6df36527210846cdaee726a9ee2d998f/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-imyh01q8
         cwd: /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-vfugjp8k/qdldl_6df36527210846cdaee726a9ee2d998f/
    Complete output (6 lines):
    usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: setup.py --help [cmd1 cmd2 ...]
       or: setup.py --help-commands
       or: setup.py cmd --help
  
    error: invalid command 'bdist_wheel'
    ----------------------------------------
    ERROR: Failed building wheel for qdldl
  ERROR: Failed to build one or more wheels
  Traceback (most recent call last):
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/installer.py", line 128, in fetch_build_egg
      subprocess.check_call(cmd)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 373, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['/Users/berk/.pyenv/versions/3.9.4/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/tmpyrlczeky', '--quiet', 'qdldl']' returned non-zero exit status 1.
  
  The above exception was the direct cause of the following exception:
  
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_a770c588033e4c6eb1f79635087d3d40/setup.py", line 261, in <module>
      setup(name='osqp',
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/__init__.py", line 164, in setup
      _install_setup_requires(attrs)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/__init__.py", line 159, in _install_setup_requires
      dist.fetch_build_eggs(dist.setup_requires)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/dist.py", line 699, in fetch_build_eggs
      resolved_dists = pkg_resources.working_set.resolve(
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 779, in resolve
      dist = best[req.key] = env.best_match(
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1064, in best_match
      return self.obtain(req, installer)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1076, in obtain
      return installer(requirement)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/dist.py", line 758, in fetch_build_egg
      return fetch_build_egg(self, req)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/installer.py", line 130, in fetch_build_egg
      raise DistutilsError(str(e)) from e
  distutils.errors.DistutilsError: Command '['/Users/berk/.pyenv/versions/3.9.4/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/tmpyrlczeky', '--quiet', 'qdldl']' returned non-zero exit status 1.
  Disabling LONG
  Remove long integers for numpy compatibility. See:
   - https://github.com/numpy/numpy/issues/5906
   - https://github.com/ContinuumIO/anaconda-issues/issues/3823
  You can reenable long integers by passing: --osqp --long argument.
  
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/c0/90/4cf48c200a89e46bcad87e12469ee36fc03d0c3f16b703b747e8c4bf618e/osqp-0.6.2.post0.tar.gz#sha256=5f0695f26a3bef0fae91254bc283fab790dcca0064bfe0f425167f9c9e8b4cbc (from https://pypi.org/simple/osqp/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
  Downloading osqp-0.6.2.tar.gz (219 kB)
     |████████████████████████████████| 219 kB 3.9 MB/s            
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_33bf7941d8d34172b2d34b2bf3c54f2e/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_33bf7941d8d34172b2d34b2bf3c54f2e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-pip-egg-info-d4iawwvm
       cwd: /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_33bf7941d8d34172b2d34b2bf3c54f2e/
  Complete output (50 lines):
  WARNING: The wheel package is not available.
    ERROR: Command errored out with exit status 1:
     command: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-v58reoup/qdldl_35260caaf25b4857b7b39d050278ce57/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-v58reoup/qdldl_35260caaf25b4857b7b39d050278ce57/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-r3s_kpza
         cwd: /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-wheel-v58reoup/qdldl_35260caaf25b4857b7b39d050278ce57/
    Complete output (6 lines):
    usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: setup.py --help [cmd1 cmd2 ...]
       or: setup.py --help-commands
       or: setup.py cmd --help
  
    error: invalid command 'bdist_wheel'
    ----------------------------------------
    ERROR: Failed building wheel for qdldl
  ERROR: Failed to build one or more wheels
  Traceback (most recent call last):
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/installer.py", line 128, in fetch_build_egg
      subprocess.check_call(cmd)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 373, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['/Users/berk/.pyenv/versions/3.9.4/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/tmp1xjv5atj', '--quiet', 'qdldl']' returned non-zero exit status 1.
  
  The above exception was the direct cause of the following exception:
  
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_33bf7941d8d34172b2d34b2bf3c54f2e/setup.py", line 261, in <module>
      setup(name='osqp',
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/__init__.py", line 164, in setup
      _install_setup_requires(attrs)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/__init__.py", line 159, in _install_setup_requires
      dist.fetch_build_eggs(dist.setup_requires)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/dist.py", line 699, in fetch_build_eggs
      resolved_dists = pkg_resources.working_set.resolve(
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 779, in resolve
      dist = best[req.key] = env.best_match(
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1064, in best_match
      return self.obtain(req, installer)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1076, in obtain
      return installer(requirement)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/dist.py", line 758, in fetch_build_egg
      return fetch_build_egg(self, req)
    File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/installer.py", line 130, in fetch_build_egg
      raise DistutilsError(str(e)) from e
  distutils.errors.DistutilsError: Command '['/Users/berk/.pyenv/versions/3.9.4/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/tmp1xjv5atj', '--quiet', 'qdldl']' returned non-zero exit status 1.
  Disabling LONG
  Remove long integers for numpy compatibility. See:
   - https://github.com/numpy/numpy/issues/5906
   - https://github.com/ContinuumIO/anaconda-issues/issues/3823
  You can reenable long integers by passing: --osqp --long argument.
  
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/7a/b8/291caf5b448de7b35969b71df83b0733c853656b0304516f6fea8f1c9f78/osqp-0.6.2.tar.gz#sha256=262162039f6ad6c9ffee658541b18cfae8240b65edbde71d9b9e3af42fbfe4b3 (from https://pypi.org/simple/osqp/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
  Downloading osqp-0.6.1.tar.gz (211 kB)
     |████████████████████████████████| 211 kB 22.7 MB/s            
  Preparing metadata (setup.py) ... done
Requirement already satisfied: ecos>=2 in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (from cvxpy) (2.0.7.post1)
Requirement already satisfied: scs>=1.1.6 in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (from cvxpy) (2.1.3)
Requirement already satisfied: numpy>=1.15 in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (from cvxpy) (1.21.4)
Requirement already satisfied: scipy>=1.1.0 in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (from cvxpy) (1.7.0)
Requirement already satisfied: future in ./.pyenv/versions/3.9.4/lib/python3.9/site-packages (from osqp>=0.4.1->cvxpy) (0.18.2)
Using legacy 'setup.py install' for cvxpy, since package 'wheel' is not installed.
Using legacy 'setup.py install' for osqp, since package 'wheel' is not installed.
Installing collected packages: osqp, cvxpy
    Running setup.py install for osqp ... error
    ERROR: Command errored out with exit status 1:
     command: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-record-rbwbjq80/install-record.txt --single-version-externally-managed --compile --install-headers /Users/berk/.pyenv/versions/3.9.4/include/python3.9/osqp
         cwd: /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/
    Complete output (127 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-12.0-arm64-3.9
    creating build/lib.macosx-12.0-arm64-3.9/osqp
    copying module/interface.py -> build/lib.macosx-12.0-arm64-3.9/osqp
    copying module/__init__.py -> build/lib.macosx-12.0-arm64-3.9/osqp
    copying module/utils.py -> build/lib.macosx-12.0-arm64-3.9/osqp
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen
    copying module/codegen/__init__.py -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen
    copying module/codegen/utils.py -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen
    copying module/codegen/code_generator.py -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen
    creating build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/mkl_pardiso_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/warm_start_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/polishing_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/codegen_matrices_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/__init__.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/update_matrices_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/primal_infeasibility_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/codegen_vectors_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/non_convex_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/unconstrained_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/dual_infeasibility_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/basic_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    copying module/tests/feasibility_test.py -> build/lib.macosx-12.0-arm64-3.9/osqp/tests
    creating build/lib.macosx-12.0-arm64-3.9/osqppurepy
    copying modulepurepy/interface.py -> build/lib.macosx-12.0-arm64-3.9/osqppurepy
    copying modulepurepy/_osqp.py -> build/lib.macosx-12.0-arm64-3.9/osqppurepy
    copying modulepurepy/__init__.py -> build/lib.macosx-12.0-arm64-3.9/osqppurepy
    running egg_info
    writing osqp.egg-info/PKG-INFO
    writing dependency_links to osqp.egg-info/dependency_links.txt
    writing requirements to osqp.egg-info/requires.txt
    writing top-level names to osqp.egg-info/top_level.txt
    listing git files failed - pretending there aren't any
    reading manifest file 'osqp.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'osqp.egg-info/SOURCES.txt'
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen/files_to_generate
    copying module/codegen/files_to_generate/CMakeLists.txt -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/files_to_generate
    copying module/codegen/files_to_generate/emosqpmodule.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/files_to_generate
    copying module/codegen/files_to_generate/example.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/files_to_generate
    copying module/codegen/files_to_generate/setup.py -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/files_to_generate
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/configure
    copying module/codegen/sources/configure/osqp_configure.h.in -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/configure
    copying module/codegen/sources/configure/qdldl_types.h.in -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/configure
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/CMakeLists.txt -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/auxil.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/constants.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/error.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/glob_opts.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/kkt.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/lin_alg.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/osqp.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/proj.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/qdldl.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/qdldl_interface.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/scaling.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/types.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    copying module/codegen/sources/include/util.h -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/include
    creating build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/CMakeLists.txt -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/auxil.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/error.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/kkt.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/lin_alg.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/osqp.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/proj.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/qdldl.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/qdldl_interface.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/scaling.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    copying module/codegen/sources/src/util.c -> build/lib.macosx-12.0-arm64-3.9/osqp/codegen/sources/src
    running build_ext
    Traceback (most recent call last):
      File "/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py", line 164, in build_extensions
        check_output(['cmake', '--version'])
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 424, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 505, in run
        with Popen(*popenargs, **kwargs) as process:
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 951, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 1821, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'cmake'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py", line 208, in <module>
        setup(name='osqp',
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/__init__.py", line 165, in setup
        return distutils.core.setup(**attrs)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/command/install.py", line 546, in run
        self.run_command('build')
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 87, in run
        _build_ext.run(self)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
        _build_ext.build_ext.run(self)
      File "/Users/berk/.pyenv/versions/3.9.4/lib/python3.9/distutils/command/build_ext.py", line 340, in run
        self.build_extensions()
      File "/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py", line 166, in build_extensions
        raise RuntimeError("CMake must be installed to build OSQP")
    RuntimeError: CMake must be installed to build OSQP
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Users/berk/.pyenv/versions/3.9.4/bin/python3.9 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py'"'"'; __file__='"'"'/private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-install-n23ks69k/osqp_869c8194dede456b85ebdf607e348803/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/j0/dcfk96ls037_51ndyw_4nymh0000gn/T/pip-record-rbwbjq80/install-record.txt --single-version-externally-managed --compile --install-headers /Users/berk/.pyenv/versions/3.9.4/include/python3.9/osqp Check the logs for full command output.
```

**Comment by SteveDiamond** [comment]
@ustunb you have the command line tools installed right?

**Comment by ustunb** [comment]
Yes that's right. 

Running `xcode-select --install` returns `xcode-select: error: command line tools are already installed, use "Software Update" to install updates`

**Comment by sunilshah** [comment]
@ustnub

In my case, I installed numpy, scipy, cmake, osqp, cvxopt, scs  before I installing cvxpy.   I used homebrew to install numpy 1.21.4.  Scipy is v 1.7.2 (v 1.7.3   slows down parallel tasks for my installation).  Dependencies such as openblas and cmake are installed using homebrew.  Most packages require compilation from source.

**Comment by ustunb** [comment]
Thank you!

I installed cmake using homebrew and it worked!

In case anyone else runs into this issue, the command that I ran was:

`pip install -U --no-use-pep517 --no-cache-dir cvxpy`

This installed cvxpy-1.1.17

I had also installed openblas via brew, and also had scipy and numpy
installed.







If anyone else runs into the same problem: I had also installed

On Thu, Dec 9, 2021 at 3:48 PM sunilshah ***@***.***> wrote:
>
> @ustnub
>
> In my case, I installed numpy, scipy, cmake, osqp, cvxopt, scs before I
installing cvxpy. I used homebrew to install numpy 1.21.4. Scipy is v 1.7.2
(v 1.7.3 slows down parallel tasks for my installation). Dependencies such
as openblas and cmake are installed using homebrew. Most packages require
compilation from source.

**Comment by andruten** [comment]
Thank you @ustunb! It worked for me. I also had to specify `ecos==2.0.7.post.1` version. See https://github.com/embotech/ecos-python/issues/33

**Comment by Shiyou999** [comment]
> Thank you @ustunb! It worked for me. I also had to specify `ecos==2.0.5` version. See [embotech/ecos-python#33](https://github.com/embotech/ecos-python/issues/33)

`pip install ecos==2.0.5` fixed my problem! Thanks a lot!

**Comment by rileyjmurray** [comment]
You should no longer need to use the ``--no-use-pep517`` flag if you're building CVXPY 1.1.19 (or newer in the 1.1.x release series) or anything in the 1.2.x release series.

I'll leave this issue open so it's easy for people to find until installation becomes more seamless.

**Comment by brettelliot** [comment]
Hi all. Im still unable to `pip install cvxpy` on my mac M1. I installed cmake with homebrew, and im using ECOS 2.0.5. Im using python 3.8.10 and here are my current packages:

ecos       2.0.5
numpy      1.22.4
osqp       0.6.2.post5
pip        22.1.2
pybind11   2.9.2
qdldl      0.1.5.post2
scipy      1.8.1
setuptools 56.0.0

The issue appears to be that despite having numpy 1.22.4, it trys to get numpy-1.16.6 and then that fails.

Any thoughts?

**Comment by phschiele** [comment]
@brettelliot There is an open PR (https://github.com/cvxpy/cvxpy/pull/1725) that would create wheels for M1. Unfortunately, we do not have an M1 machine to test this at the moment.
You can give us feedback if downloading the wheel file from the bottom of the build (https://github.com/cvxpy/cvxpy/actions/runs/2029620352) would work for you.
The corresponding file from the zip is called "cvxpy-1.3.0-cp38-cp38-macosx_10_9_universal2.whl".
After extracting, you should be able to simply 
```
pip install cvxpy-1.3.0-cp38-cp38-macosx_10_9_universal2.whl
```

**Comment by brettelliot** [comment]
Hi @phschiele! That worked! Thanks!

**Comment by rileyjmurray** [comment]
@brettelliot @phschiele I'll caution that the version number there is dubious! We haven't released CVXPY 1.3.0 yet, on any platform. You're getting a development version of some kind by installing that (something after 1.2.0, but probably before 1.2.1).

**Comment by phschiele** [comment]
@rileyjmurray Thanks, for the word of caution here!
For reference, the linked PR is based on this commit in master: https://github.com/cvxpy/cvxpy/commit/05157b0f6b38e736918970ee6b0d51d27793dccf

Do you suggest that M1 wheels would become available with 1.3?
Otherwise, we could merge the PR also in the release branches and rerun the deploy to backfill the wheels for 1.1.x and 1.2.x.

**Comment by rileyjmurray** [comment]
I think merging this PR in the release branches to backfill wheels is perfectly fine. We'd just be changing infrastructure, not any cvxpy code that a user gets. 

I specifically think we can backport wheels for the most recent releases on 1.2.x (i.e., 1.2.1) and 1.1.x (not sure of the precise number).

**Comment by andruten** [comment]
> @brettelliot There is an open PR (#1725) that would create wheels for M1. Unfortunately, we do not have an M1 machine to test this at the moment. You can give us feedback if downloading the wheel file from the bottom of the build (https://github.com/cvxpy/cvxpy/actions/runs/2029620352) would work for you. The corresponding file from the zip is called "cvxpy-1.3.0-cp38-cp38-macosx_10_9_universal2.whl". After extracting, you should be able to simply
> 
> ```
> pip install cvxpy-1.3.0-cp38-cp38-macosx_10_9_universal2.whl
> ```

I can confirm cvxpy is installed well in a M1 having python 3.9 too :). Thank you!

**Comment by SteveDiamond** [comment]
Resolved via @phschiele 's PR.

**Comment by Estebang17-a** [comment]
Hello, I can't download the wheel as the build expired. Could you help me please?
